### ISSUE TITLE
feature request: Allow moving file into a new directory when the name contains a slash 

### ISSUE THREAD

**Issue by titouancreach** [stevearc/oil.nvim#117]
### Did you check existing requests?

- [X] I have searched the existing issues

### Describe the feature

I have a file like:

foobar.js

if if rename the file to 

hello/foobar.js

I'd like to create a directory `hello` and put `foobar.js` in it

### Additional details

If it's a feature you may like, i can write a PR

**Comment by stevearc** [comment]
This is a neat idea, but it's more complicated than it first seems. The main issue is the possibility that `hello/foobar.js` or even just `hello/` already exist. It's doable, but I'll have to be real careful about the edge cases.

I'll keep this on the back burner until I get the time to look deeper into it.

**Comment by rasulomaroff** [comment]
+1 for this one!

I've been using `oil.nvim` for quite a while already and absolutely love it! I think this feature will kind of complete all the flow that typically happens when editing/creating files/folders.

**Comment by NReilingh** [comment]
Another fun complication :) -- move file to parent directory by renaming it `../file.txt` or `./../file.txt`. For some reason this (as well as subdirectories) feels very intuitive for me in the context of the rest of oil.nvim's features.

**Comment by rasulomaroff** [comment]
> Another fun complication :) -- move file to parent directory by renaming it `../file.txt` or `./../file.txt`. For some reason this (as well as subdirectories) feels very intuitive for me in the context of the rest of oil.nvim's features.

Absolutely! Or moving it into a directory inside a parent one by `../dir/file.txt`

**Comment by titouancreach** [comment]
another one (for fun): transforming a file to a folder if the file is empty and without extension. Use case: I create a foobar file: hoo noo, i wanter to create a folder but a created an extensionless file instead, rename to foobar/

**Comment by rasulomaroff** [comment]
> if the file is empty and without extension

I think that doesn't really necessary. Directories with dots are perfectly valid. Alternatively, if a file is not empty and you are transforming it into a directory, then this information can easily be shown in the confirmation window.

But I would go with any way @stevearc decides to implement, if he does.

**Comment by stevearc** [comment]
@titouancreach congratulations, your pain is also my pain! And that's a much easier fix, so I've gone ahead and implemented it. Now you can change a file to a dir or vice-versa by adding/removing the trailing `/`. It'll translate that change to a delete + create.

For the other requests in this issue, I'm still looking into it. I have some ideas and a rough plan now, but I just haven't had the unbroken time to sit down and hash it out.

**Comment by titouancreach** [comment]
Great job @stevearc ! I update my packages right now !

**Comment by LittleTealeaf** [comment]
Just discovered Oil and love it so far, but this is definitely one of the things I'm hoping gets added. Currently my solution is to continue using telescope file browser to handle moving between directories, but I'd definitely LOVE to be able to do that in Oil :)

**Comment by augustdolan** [comment]
Hey pal! Love the plugin a ton. Totally heard on the complexities involved in this feature request, too. I'm here because I also just attempted to do this intuitive feeling ask, and the operation of course failed. 

Let me know if there's anything I can do to help you with this feature. Thanks for an amazing project!

**Comment by danrasmuson** [comment]
Loving this plugin and I'm feeling the same way as those above. Would be a great addition to this awesome package.

**Comment by pidgeon777** [comment]
Hello, sorry but I have not understood the whole discussion, I would like to kindly ask if it currently possible to automatically create missing folders during file operation.

If still not possible, I agree this would be a must-have feature (if feasible).

**Comment by pi314ever** [comment]
Made a PR which I think works for all the cases I can think of. Let me know if I missed anything!

**Comment by danrasmuson** [comment]
This is working great for me! Thank you!

**Comment by ruigouveiamaciel** [comment]
moving files into the parent directory using `../file_name.txt` still doesn't work, are there any plans for implementing this? I really enjoy oil but being able to move things to the parent directory is quite important to me when I am refactoring things. For now I am gonna have to use @pi314ever fork as that feature works there.

**Comment by swahpy** [comment]
Hi, moving a file to an existed folder doesn't work for me, either. Just as below, I want to move `auto-session.lua` to `archive/auto-session.lua`, but it prompted that `Filename cannot contain path separator`. if I rewrite the whole line to `archive/auto-session.lua` regardless of the icon at the beginning then it will just create a blank new file istead of moving file. 
I really like oil and reply on it heavily. Please correct me if I've got it wrong. If the feature does miss for the moment, hope that we could consider to add it in the near future. Thank you in advance.
![image](https://github.com/stevearc/oil.nvim/assets/57909991/aa2b72ef-c6d1-4bb5-9356-6cf21780f3c1)

**Comment by oxi1224** [comment]
> Hi, moving a file to an existed folder doesn't work for me, either. Just as below, I want to move `auto-session.lua` to `archive/auto-session.lua`, but it prompted that `Filename cannot contain path separator`. if I rewrite the whole line to `archive/auto-session.lua` regardless of the icon at the beginning then it will just create a blank new file istead of moving file. I really like oil and reply on it heavily. Please correct me if I've got it wrong. If the feature does miss for the moment, hope that we could consider to add it in the near future. Thank you in advance. ![image](https://private-user-images.githubusercontent.com/57909991/333625877-aa2b72ef-c6d1-4bb5-9356-6cf21780f3c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0Mzk4MTEsIm5iZiI6MTcyMDQzOTUxMSwicGF0aCI6Ii81NzkwOTk5MS8zMzM2MjU4NzctYWEyYjcyZWYtYzZkMS00YmI1LTkzNTYtNmNmMjE3ODBmM2MxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA4VDExNTE1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzNDQzYzQyMmM5YzA0YzRmNWVkMTYzMzg1MGExOGVmZTgzODg0ZTkwNzVhZDE2Nzc2YzA4NTEzNmZlMDI2ZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AZwHFgLl42XbTon7HG1hSNMqDL-iart3iHG9vLVfAMg)

Yep, same thing still happening

**Comment by ruigouveiamaciel** [comment]
For anyone still having this issue, my suggestion is to simply delete the file you want to use with `dd` (do not save the deletion), then move into the folder you want to move the file into, paste the deleted file and then Save. It will prompt you to move the file.

**Comment by akiross** [comment]
> This is a neat idea, but it's more complicated than it first seems. The main issue is the possibility that `hello/foobar.js` or even just `hello/` already exist. It's doable, but I'll have to be real careful about the edge cases.

I think it might be useful to cover the most simple and unambiguous case first: if the folder does not exists and there are no duplicated files when trying to move, then it can be created and al the files moved at once. It seems "future-proof" enough, so more complex edge cases might be treated in the future. In my case, this would be enough most of the times.

**Comment by ndavd** [comment]
> I think it might be useful to cover the most simple and unambiguous case first: if the folder does not exists and there are no duplicated files when trying to move, then it can be created and al the files moved at once. It seems "future-proof" enough, so more complex edge cases might be treated in the future. In my case, this would be enough most of the times.

I agree. That would greatly improve its usability.
